packages/core/src/storage.ts

```diff
@@ -136,28 +168,171 @@ export function describeError(e: StorageOperation): string {
+    return {
+      ...item,
+      type: addDetailsToParsedType(parseTypeId(item.type)),
+    };
+  }
+
+  function addDetailsToParsedType(parsed: ParsedTypeId): ParsedTypeDetailed {
+    const item = layout.types[parsed.id];
+    const members =
+      item?.members && (isStructMembers(item?.members) ? item.members.map(parseWithDetails) : item?.members);
+    return {
+      ...parsed,
+      args: parsed.args?.map(addDetailsToParsedType),
+      rets: parsed.args?.map(addDetailsToParsedType),
+      item: { ...item, members },
+    };
+  }
+
+  return layout.storage.map(parseWithDetails);
+}
+
+function isEnumMembers<T>(members: TypeItemMembers<T>): members is EnumMember[] {
+  return members.length === 0 || typeof members[0] === 'string';
+}
+
+function isStructMembers<T>(members: TypeItemMembers<T>): members is StructMember<T>[] {
+  return members.length === 0 || typeof members[0] === 'object';
+}
+
+function matchStorageField(original: StorageField, updated: StorageField, { canGrow }: { canGrow: boolean }) {
+  const nameMatches = original.label === updated.label;
+  const typeMatches = compatibleTypes(original.type, updated.type, { canGrow });
+
+  if (typeMatches && nameMatches) {
+    return 'equal';
+  } else if (typeMatches) {
+    return 'rename';
+  } else if (nameMatches) {
+    return 'typechange';
+  } else {
+    return 'replace';
+  }
+}
+
+// TODO: Avoid getting stuck with recursive types.
+// struct S { mapping (uint => S) s; }
+
+function compatibleTypes(
+  original: ParsedTypeDetailed,
+  updated: ParsedTypeDetailed,
+  { canGrow }: { canGrow: boolean },
+): boolean {
+  if (original.head !== updated.head) {
+    return false;
+  }
+
+  if (original.args === undefined || updated.args === undefined) {
+    // both should be undefined at the same time
+    assert(original.args === updated.args);
+    return true;
+  }
+
+  const { head } = original;
+
+  if (head === 't_contract') {
+    // no storage layout errors can be introduced here since it is just an address
+    return true;
+  }
+
+  if (head === 't_struct') {
+    const originalMembers = original.item.members;
+    const updatedMembers = updated.item.members;
+    assert(originalMembers && isStructMembers(originalMembers));
+    assert(updatedMembers && isStructMembers(updatedMembers));
+    const errors = getStorageUpgradeErrorsGeneric(originalMembers, updatedMembers, { canGrow });
+    return errors.length === 0;
+  }
+
+  if (head === 't_enum') {
+    const originalMembers = original.item.members;
+    const updatedMembers = updated.item.members;
+    assert(originalMembers && isEnumMembers(originalMembers));
+    assert(updatedMembers && isEnumMembers(updatedMembers));
+    const ops = levenshtein(originalMembers, updatedMembers, (a, b) => (a === b ? 'equal' : 'replace'));
+    // it is only allowed to append new enum members
+    return ops.every(o => o.kind === 'append');
```

Comment:
Wasn't there an issue with a different underlying type being used for enums depending on how many possible values it has? Eg enums are backed by uint8s as long as they fit, but if you have too many members it may switch to a uint16. If that's the case, we should guard against appends that change the type of the backing field.

Reply:
Yes, in fact newer Solidity versions will limit an enum to 256 elements to avoid this issue. But we should add this check for older versions. I would limit to 256 maximum elements to ensure compatbility the newer Solidity versions.